discovery_node/substrate-ingest/src/joynode/QueryNode.ts

```diff
@@ -0,0 +1,95 @@
+// @ts-check
+import { ApiPromise, WsProvider } from '@polkadot/api';
+import { registerJoystreamTypes } from '@joystream/types';
+import Config from '../Config'
+import { State } from '../state/StateKeeper'
+import ISubstrateQueryService, { makeQueryService } from './ISubstrateQueryService';
+import QueryBlockProducer from './QueryBlockProducer';
+import QueryEventBlock from './QueryEventBlock';
+import { EventEmitter } from 'events';
+import StateBootstrap from './StateBootstrap';
+
+const logger = require('log4js').getLogger('query_node');
+
+export enum QueryNodeState {
+    NOT_STARTED,
+    STARTING,
+    STARTED,
+    STOPPING,
+    STOPPED,
+  }
+
+export default class QueryNode extends EventEmitter {
+    private _state: QueryNodeState;
+
+    private _websocketProvider?: WsProvider;
+    
+    // API instance for talking to Substrate full node.
+    private _queryService?: ISubstrateQueryService;
+    private _queryBlockProducer?: QueryBlockProducer;
+    private _stateBootstrap?: StateBootstrap;
+    private _api?: ApiPromise;
+
+    // Query index building node.
+    //private _indexBuilder: IndexBuilder;
+
+    constructor() {
+        super();
+        this._state = QueryNodeState.NOT_STARTED;
+    }
+
+    async build(config: Config) {
+        let providerUrl = config.get()?.joystream?.ws_provider || process.env.JOYSTREAM_PROVIDER;
+        if (!providerUrl) {
+            throw Error('Invalid config: joystream.ws_provider is not set');
+        }
+
+        this._websocketProvider = new WsProvider(providerUrl);
+        registerJoystreamTypes();
+        
+        this._api  = await ApiPromise.create({ provider: this._websocketProvider });
+        this._queryService = makeQueryService(this._api);
+        this._queryBlockProducer = new QueryBlockProducer(this._queryService, config);
+        this._stateBootstrap = new StateBootstrap(this._queryService, config);
+    }
+
+    async start(state: State) {
+        if (!this._queryBlockProducer) {
+            throw Error('Query block producer is not initialized');
+        }
+
+        //this._queryBlockProducer.on('QueryEventBlock', (query_event_block: QueryEventBlock): void => {
+        //    this._onQueryEventBlock(query_event_block);
+        //});
+
+        this._state = QueryNodeState.STARTED;
+        logger.debug(`Starting at state ${state}`);
+        await this._queryBlockProducer.start(state.lastProcessedBlock + 1);
```

Comment:
Agree, `StateKeeper` is thought to be responsible for the fail-safe behaviour. In the recovery mode, it will restart the producer from the last fully processed blocks and simply skip the already processed events in the next block if necessary.